keylime/cert_utils.py

```diff
@@ -1,13 +1,133 @@
-from cryptography.hazmat.primitives.serialization import load_der_public_key
+import io
+import os.path
+import subprocess
+import sys
+
+from cryptography import exceptions as crypto_exceptions
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import ec, padding
+from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPublicKey
 from pyasn1.codec.der import decoder, encoder
-from pyasn1_modules import rfc2459
+from pyasn1_modules import pem, rfc2459
 
+from keylime import config, keylime_logging, tpm_ek_ca
 
 # Issue #944 -- python-cryptography won't parse malformed certs,
 # such as some Nuvoton ones we have encountered in the field.
 # Unfortunately, we still have to deal with such certs anyway.
-# Let's read the EK cert with pyasn1 instead of python-cryptography.
-def read_x509_der_cert_pubkey(der_cert_data):
-    """Returns the public key of a DER-encoded X.509 certificate"""
-    der509 = decoder.decode(der_cert_data, asn1Spec=rfc2459.Certificate())[0]
-    return load_der_public_key(encoder.encode(der509["tbsCertificate"]["subjectPublicKeyInfo"]))
+
+# Here we provide some helpers that use pyasn1 to parse the certificates
+# when parsing them with python-cryptography fails, and in this case, we
+# try to read the parsed certificate again into python-cryptograhy.
+
+logger = keylime_logging.init_logging("cert_utils")
+
+
+class UserError(Exception):
+    pass
+
+
+def x509_der_cert(der_cert_data):
+    if isinstance(der_cert_data, str):
+        der_cert_data = der_cert_data.encode("utf-8")
+    try:
+        cert = cert = x509.load_der_x509_certificate(data=der_cert_data, backend=default_backend())
+        return cert
+    except Exception:
+        pyasn1_cert = decoder.decode(der_cert_data, asn1Spec=rfc2459.Certificate())[0]
+        return x509.load_der_x509_certificate(data=encoder.encode(pyasn1_cert), backend=default_backend())
+
+
+def x509_pem_cert(pem_cert_data):
+    if isinstance(pem_cert_data, (bytes, bytearray)):
+        pem_cert_data = pem_cert_data.decode("utf-8")
+    try:
+        cert = x509.load_pem_x509_certificate(data=pem_cert_data, backend=default_backend())
+        return cert
+    except Exception:
+        return x509_der_cert(pem.readPemFromFile(io.StringIO(pem_cert_data)))
```

Comment:
That confuses me, if is a call to read a PEM, why do we try a DER here? This make the operation non-symmetric (a call to read a PEM can try a DER, but no vice versa)

If symmetry is required, IMHO is a bit more clear if:
* The DER function try the pyasm1 decoder after the conversion to bytes, capturing the `pyasn1.error.SubstrateUnderrunError` exception to see if it valid or not, and calling the load_der_... only one
* Drop in the DER function the conversion to bytes (see later)
* In the PEM function only try the load_pem_ ... call
* Drop in the PEM function the conversion from bytes
* Create a new `x509_cert` that order the calls to PEM / DER in the order that you suit, and do the conversion from / to bytes
* Use only this function to load the certificates, and never call the PEM / DER versions outside this new function

Reply:
PEM here is the DER data base64 encoded, with the added header and footer (e.g. `----BEGIN CERT...`), so once we call `pem.readPemFromFile()` we have the actual bytes, which we can parse with the method that parses the DER data. I changed it slightly and added a comment, to try to clarify it a bit, but I am not sure I addressed your concern fully, so if you could recheck I would appreciate.